### PR TITLE
Convert web application to TypeScript

### PR DESCRIPTION
**Pull request by ashmind**
TODO:
- [x] Finish build work
- [x] Convert build.js to experimental-modules
- [x] Consolidate .eslintrc.json files

### CODE REVIEW

File: source/WebApp/.eslintrc.json [hunk 1]
```diff
@@ -1,43 +1,35 @@
...
     },
     "globals": {
         "Promise": true,
         "Symbol": true
     },
-    "parserOptions": {
-        "ecmaVersion": 8
-    },
     "rules": {
-        "no-cond-assign": "error",
```
Comment (jsk95) [inline source/WebApp/.eslintrc.json:14]: I like how gradually less custom you are getting :)

File: source/WebApp/build.js [hunk 2]
```diff
@@ -69,32 +64,19 @@ task('less', async () => {
...
-            }),
-            {
-                name: 'rollup-plugin-adhoc-resolve-vue',
-                resolveId: id => (id === 'vue') ? path.resolve(`./node_modules/vue/dist/vue${production?'.min':''}.js`) : null
-            },
-            rollupPluginNodeResolve({ browser: true }),
-            ...(production ? [rollupPluginTerser()] : [])
-        ]
-    });
+task('tsLint', () => execa.command('eslint . --max-warnings 0 --ext .js,.ts', {
```
Comment (jsk95) [inline source/WebApp/build.js:67]: should be eslint or just lint

File: source/WebApp/components/internal/codemirror/addon-jump-arrows.ts [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+        TElement extends SVGCircleElement ? { cx: string|number; cy: string|number; r: string|number } : {}
+    )
+) {
+    for (const key in attributes) {
+        element.setAttribute(key, attributes[key as keyof typeof attributes] as string);
+    }
+}
+
+class ArrowLayer {
+    private root: SVGSVGElement;
```
Comment (jsk95) [inline source/WebApp/components/internal/codemirror/addon-jump-arrows.ts:64]: Think some of these could be `readonly`

File: source/WebApp/components/internal/codemirror/addon-jump-arrows.ts [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+interface JumpArrow {
+    readonly key: string;
+    readonly from: Readonly<CodeMirror.Position>;
+    readonly to: Readonly<CodeMirror.Position>;
+    readonly bookmarks: JumpBookmarks;
+    readonly options: JumpOptions;
+    readonly svg: JumpSvg;
+}
+
+function debounce(func: () => void, interval: number) {
```
Comment (jsk95) [inline source/WebApp/components/internal/codemirror/addon-jump-arrows.ts:34]: Think I literally just saw a debounce import 
`import { debounce } from 'throttle-debounce';`

Are they different?

File: source/WebApp/components/internal/codemirror/mode-asm.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
+// CodeMirror, copyright (c) by Marijn Haverbeke and others
+// Distributed under an MIT license: http://codemirror.net/LICENSE
+import CodeMirror from 'codemirror';
+
+CodeMirror.defineMode('asm', () => {
+    'use strict';
```
Comment (jsk95) [inline source/WebApp/components/internal/codemirror/mode-asm.ts:6]: Why do you need `use strict` here?

File: source/WebApp/components/internal/graph-layout/force-bind-nested.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import type { ExtendedNodeDatum, NestedNodeDatumData } from './interfaces';
+
+export default () => {
+    let nodes: ReadonlyArray<ExtendedNodeDatum<NestedNodeDatumData>>;
+    const force = () => {
+        for (const node of nodes) {
+            const { nested: { parent, dx, dy } } = node.data;
```
Comment (jsk95) [inline source/WebApp/components/internal/graph-layout/force-bind-nested.ts:7]: `const { parent, dx, dy } = node.data.nested;`

File: source/WebApp/package.json [hunk 7]
```diff
@@ -2,44 +2,63 @@
   "name": "sharplab",
   "version": "0.0.0",
   "private": true,
+  "type": "module",
   "scripts": {
-    "build": "node build.js",
-    "watch": "node build.js --watch",
-    "lint": "eslint . --ignore-path ../../.gitignore --max-warnings 0",
+    "build": "check-node-version --node \">=12.6.1\" && node --experimental-modules build.js",
```
Comment (jsk95) [inline source/WebApp/package.json:7]: https://stackoverflow.com/a/50798589

File: source/WebApp/ts/helpers/throw-error.ts [hunk 8]
```diff
@@ -0,0 +1,3 @@
+export default function throwError(message: string): never {
```
Comment (jsk95) [inline source/WebApp/ts/helpers/throw-error.ts:1]: :hmm:

File: source/WebApp/ts/state/handlers/url/load-gist-async.ts [hunk 9]
```diff
@@ -1,15 +1,22 @@
-import { getGistAsync } from '../../../helpers/github/gists.js';
-import languages from '../../../helpers/languages.js';
-import { languageAndTargetMapReverse } from '../helpers/language-and-target-maps.js';
+import type { TargetName } from '../../../helpers/targets';
```
Comment (jsk95) [inline source/WebApp/ts/state/handlers/url/load-gist-async.ts:1]: TIL

File: source/WebApp/tests/rendering/render.ts [hunk 10]
```diff
@@ -80,6 +80,11 @@ export default async function render({
     styles = [],
 
     debug = false
+}: {
+    html: string;
+    bodyClass?: string;
+    styles?: ReadonlyArray<{ path: string }>;
+    debug?: boolean;
```
Comment (ashmind) [inline source/WebApp/tests/rendering/render.ts:87]: All `readonly`?

File: source/WebApp/ts/state/handlers/url/load-gist-async.ts (hunk 9 above)
Comment (ashmind) [reply source/WebApp/ts/state/handlers/url/load-gist-async.ts:1]: Need to add `"importsNotUsedAsValues": "error"` at work.

File: source/WebApp/ts/types/gist.ts [hunk 11]
```diff
@@ -0,0 +1,9 @@
+import type { AppOptions } from './app';
+
+export interface Gist {
+    id: string;
+    name: string;
+    url: string;
+    options: AppOptions;
+    code: string;
```
Comment (ashmind) [inline source/WebApp/ts/types/gist.ts:8]: readonly?

File: source/WebApp/ts/types/highlighted-range.ts [hunk 12]
```diff
@@ -0,0 +1,7 @@
+export type HighlightedRange = {
+    start: CodeMirror.Position;
+    end: CodeMirror.Position;
+}|{
+    start: number;
+    end: number;
+};
```
Comment (ashmind) [inline source/WebApp/ts/types/highlighted-range.ts:7]: readonly

File: source/WebApp/ts/types/code-range.ts [hunk 13]
```diff
@@ -0,0 +1,4 @@
+export interface CodeRange {
+    start: CodeMirror.Position;
+    end: CodeMirror.Position;
+}
```
Comment (ashmind) [inline source/WebApp/ts/types/code-range.ts:4]: readonly?

File: source/WebApp/ts/state/theme.ts [hunk 14]
```diff
@@ -0,0 +1,9 @@
+export type Theme = 'light'|'dark'|'auto';
```
Comment (ashmind) [inline source/WebApp/ts/state/theme.ts:1]: Should probably move to App.

File: source/WebApp/ts/helpers/throw-error.ts (hunk 8 above)
Comment (ashmind) [reply source/WebApp/ts/helpers/throw-error.ts:1]: https://github.com/tc39/proposal-throw-expressions

**Pull request merged**